core/App/ehsm_napi.cpp

```diff
@@ -1139,83 +1140,549 @@ char* NAPI_ExportDataKey(const char* cmk_base64,
+        req_msg2->kdf_id = msg2_json.readData_uint16("kdf_id");
+
+        // // sign_gb_ga
+        json_key.clear();
+        json_key = json_key + "sign_gb_ga" + LAYERED_CHARACTER + "x";
+        msg2_json.readData_uint32Array(json_key, req_msg2->sign_gb_ga.x);
+        json_key.clear();
+        json_key = json_key + "sign_gb_ga" + LAYERED_CHARACTER + "y";
+        msg2_json.readData_uint32Array(json_key, req_msg2->sign_gb_ga.y);
+
+        // mac
+        msg2_json.readData_uint8Array("mac",req_msg2->mac);
+
+        // sig_rl_size
+        req_msg2->sig_rl_size = msg2_json.readData_uint32("sig_rl_size");
+
+        // sig_rl
+        msg2_json.readData_uint8Array("sig_rl",req_msg2->sig_rl);
+
+        // setting return msg2_size
+        *msg2_size = sizeof(sgx_ra_msg2_t) + req_msg2->sig_rl_size;
+    }
+    else
+    {
+        return EH_ARGUMENTS_BAD;
+    }
+    return EH_OK;
+}
+
+/*
+ * process will be return to Enroll APP msg3 json
+ *  @param p_msg3 : will be parse sgx_ra_msg3_t
+ *  @param quote_size : quote size
+ *  @param retJsonObj : return msg3 json object 
+ *  retJsonObj : RetJsonObj
+ *      {
+ *          "mac" : array(int),
+ *          "g_a" : Json::Value
+ *              {
+ *                  gx : array(int),
+ *                  gy : array(int)
+ *              },
+ *          "ps_sec_prop" : Json::Value
+ *              {
+ *                  sgx_ps_sec_prop_desc : array(int)
+ *              },
+ *          "quote_size" : int,
+ *          "quote" : array(int)
+ *      }
+ */
+ehsm_status_t ra_proc_msg3(sgx_ra_msg3_t *p_msg3, RetJsonObj *retJsonObj, uint32_t quote_size)
+{
+    ehsm_status_t ret = EH_OK;
+
+    if(p_msg3 == NULL || retJsonObj == NULL)
+    {
+        return EH_ARGUMENTS_BAD;
+    }
+
+    JsonObj msg3_json;
+
+    std::string json_key; // json key use for string connect
+    // mac
+    msg3_json.addData_uint8Array("mac", p_msg3->mac, SGX_MAC_SIZE);
+
+    // g_a
+    json_key.clear();
+    json_key = json_key + "g_a" + LAYERED_CHARACTER + "gx";
+    msg3_json.addData_uint8Array(json_key, p_msg3->g_a.gx, SGX_ECP256_KEY_SIZE);
+    json_key.clear();
+    json_key = json_key + "g_a" + LAYERED_CHARACTER + "gy";
+    msg3_json.addData_uint8Array(json_key, p_msg3->g_a.gy, SGX_ECP256_KEY_SIZE);
+
+    // ps_sec_prop
+    json_key.clear();
+    json_key = json_key + "ps_sec_prop" + LAYERED_CHARACTER + "sgx_ps_sec_prop_desc";
+    msg3_json.addData_uint8Array(json_key, p_msg3->ps_sec_prop.sgx_ps_sec_prop_desc, 256);
+
+    // quote_size
+    msg3_json.addData_uint32("quote_size", quote_size);
+
+    // quote
+    msg3_json.addData_uint8Array("quote", p_msg3->quote, quote_size);
+
+    // setting return msg3_json
+    retJsonObj->setResult(msg3_json);
+
+    return ret;
+}
+
+/*
+ * process receive msg4 json
+ *  @param ra_msg4 : receive msg4 json string
+ *  @param req_msg4 : return sample_ra_att_result_msg_t
+ */
+ehsm_status_t ra_proc_msg4(std::string ra_msg4, sample_ra_att_result_msg_t *req_msg4)
+{
```

Comment:
1. static.
2. change to ra_proc_att_result_msg()

Reply:
The function will be move to marshal.cpp and changed name.